crud/common/sharding_key.lua

```diff
@@ -115,6 +116,87 @@ function sharding_key_module.fetch_on_router(space_name)
     return cache.sharding_key_def
 end
 
+-- Make sure sharding key definition is a part of primary key.
+function sharding_key_module.is_part_of_pk(space_name, space_format,
+                                           primary_index, sharding_key_def)
+    dev_checks('string', 'table', 'table', 'table')
+
+    if cache.as_index_object[space_name] == nil then
+        cache.as_index_object[space_name] =
+            sharding_key_module.as_index_object(space_format, space_name,
+                                                sharding_key_def)
```

Comment:
Is there a reason to cache it there, not in the `sharding_key_module.as_index_object()` function?

Reply:
Moved it back to `extract_from_pk()` function.

Reply:
BTW the reason was make it possible to test function in a unit test.
I don't know how to test a function `is_part_of_pk()` when it take sharding key as an index object implicitly (from a cache).

Reply:
Solved it.

```lua
--- a/crud/common/sharding_key.lua
+++ b/crud/common/sharding_key.lua
@@ -119,14 +119,16 @@ function sharding_key_module.update_sharding_keys_cache(space_name)
 end
 
 -- Make sure sharding key definition is a part of primary key.
-local function is_part_of_pk(space_name, primary_index)
-    dev_checks('string', 'table')
+-- sharding_key_as_index_obj is not used in CRUD, it was added to be able to
+-- use it in unit tests.
+local function is_part_of_pk(space_name, primary_index, sharding_key_as_index_obj)
+    dev_checks('string', 'table', 'table|nil')
 
     if cache.is_part_of_pk[space_name] ~= nil then
         return cache.is_part_of_pk[space_name]
     end
 
-    local sharding_key_as_index_obj = cache.as_index_object[space_name]
+    local sharding_key_as_index_obj = sharding_key_as_index_obj or cache.as_index_object[space_name]
     local is_part_of_pk = true
     local pk_fieldno_map = utils.get_index_fieldno_map(primary_index.parts)
     for _, part in ipairs(sharding_key_as_index_obj.parts) do
```